### ISSUE TITLE
memdb.MemKv put and query performance.

### ISSUE THREAD

**Issue by convexstrictly** [xtdb/xtdb#309]
Crux looks exciting.  Thanks for open sourcing it.
```clojure
(def ^crux.api.ICruxAPI system
  (crux/start-standalone-system {:kv-backend "crux.kv.memdb.MemKv"
                                 :db-dir "data/db-dir-1"
                                 :event-log-dir "data/eventlog-1"}))
(time
 (let [ns (-> 2 (math/expt 14) range)]
   (crux/submit-tx
    system
    (vec
     (for [n ns]
       [:crux.tx/put
        {:crux.db/id (keyword (str n))
         :value n
         :next (-> n inc str keyword)}]
       )))))
"Elapsed time: 2312.620042 msecs"

(time
 (let [db (crux/db system)]
   (dotimes [n 1024]
     (crux/q db
             '{:find [e1] 
               :where [[e1 :value 43]]}))))
"Elapsed time: 422.55999 msecs"

(time
 (let [db (crux/db system)]
   (dotimes [n 1024]
     (crux/q db
             '{:find [e1 e2 e3 e4 e5] 
               :where [[e1 :next e2]
                       [e2 :next e3]
                       [e3 :next e4]
                       [e4 :next e5]
                       [e5 :value 43]]}))))
"Elapsed time: 1257.500458 msecs"

```
Kudos for writing a good query optimizer.  I am seeing orders of magnitude speedup over DataScript.

Questions:
- Why is the query overhead so high for the simple query of a record by value.  In theory a single execution of looking up a record by value in MemKv should be in the 10s or 100s of nanoseconds.  Does your design preclude that kind of high performance for in-memory data, which would also cover the common case where the data is persisted on disk but has already made it to memory cache?
- If that kind of high performance is possible, better documentation or code walkthroughs of the internals may help external contributors make the necessary changes.
- Is there a way to prepare/compile a query so that the query planning costs are paid only once?  I have seen some mention in the issues about it being done automatically.
- Do you support parametrized prepared queries?
- Higher data import speed would be nice and may be necessary for certain use-cases.
- I'm surprised crux/submit-tx won't take a lazy sequence but instead insists on a vector.  For large imports that may be a problem.

I did look through the docs, but didn't see these questions answered. Sorry if I missed them.  Please let me know if you want me to open separate issues for some of these questions.

Overall very impressed by the product vision and execution so far.  Thanks again.

**Comment by refset** [comment]
Hi - thank you for the kind words and detailed feedback! We will respond more fully as soon as we can, but in the meantime...

1. Query overhead - I'm not clear what the overheads involved are, but I know the current query engine has not undergone any serious performance tuning to make trivial queries faster. Complex queries have been the priority. For related discussion see the open issue https://github.com/juxt/crux/issues/287 (and the now-closed issue linked within)
2. Internal architecture docs - We would be very glad to see other people dive into the internals of the query engine to contribute changes, and I agree good documentation there would definitely help. Unfortunately we've not been focused in this area so far, but we would be happy to give you some 1-on-1 guidance if you are particularly keen in the meantime.
3. Pre-compiled queries - queries are currently compiled and cached automatically. I'm not sure what benefits could be easily gained by adding manual control options but we will have a think.
4. Parameterised compiled queries - I can't see that we do provide this capability currently. I believe `:args` are compiled into the query. It seems to me like a reasonable requirement, and maybe there's an easy solution here. Again, we'll have a think.
5. Faster data import - this is definitely on the backlog for benchmarks, and we've not yet attempted much optimisation here.
6. Lazy seqs for `submit-tx` - we've had this come up in discussions elsewhere already and we haven't yet conclusively decided against the idea but I'm now wondering whether a higher-level API which provides batching and accepts lazy seqs is a better answer. Please feel very welcome to follow along and discuss here https://github.com/juxt/crux/issues/285

Keep it coming :)

**Comment by hraberg** [comment]
@refset captures the gist of it, just adding a few reflections here as well.

1-3. See above.

4. `:args` values are not part of the cache key, so different arguments can reuse the compiled query. 

5. The internal event log the standalone node uses hasn't been tuned. It might be replaced with the JDBC version backed by H2 or SQLite, see #270. There is also overhead unrelated to the event log implementation itself, like serialisation, spec checking and coercion.

6. As @refset mentions, we decided against supporting lazy seqs at the API level, as batching, error handling and other concerns might be better dealt with by the code doing the ingest. Individual transactions are assumed to not be "large", as they have to fit a single Kafka message. The size of this can be configured via `max.request.size`. It's possible to transact many transactions against the same valid time if necessary when batching for some consistency.

**Comment by hraberg** [comment]
I'm closing this as it's a quite wide ranging issue and more a set of questions. It's not that I disagree that we shouldn't work or fix these things, and some of it is covered and on our road map. Of course, feel free to open new individual issues!
